In [1]:
import numpy as np

Here we have a nueron which looks like a 2D version of the Sputnik satellite. The inputs are our data (famous numbers btw, extra credit if you know what they are (**without** looking them up), Weights are the things we 'tweak', and Bias is another number we can tweak to get the desired output.

![NEURON](neuron.png) 

In [2]:
inputs = [3.141, 2.1718, 1.618]
weights = [.32, .37, .62]
bias = 1.38

The inputs map directly to each weight.

And doing some simple elementry school multiplication and addition we get an output. We're skipping the Activation function for now.

In [3]:
output = (inputs[0]*weights[0]+inputs[1]*weights[1]+inputs[2]*weights[2] + bias)
print(output)

4.191846


Good for something simple like this, but scale it up to hundreds of millions/billions, that's way too much typing.

So, NumPy and Linear Algebra to the rescue!

Python has lists like our inputs. Linear Algebra also has lists but name them vectors instead. But there is a difference between a row vector and a column vector. A row vector is a 1xN matrix (a 1-dimensional array of numbers) [3.141, 2.1718, 1.618]. A column vector is an Nx1 matrix (a list of numbers arranged vertically).\
3.141 \
2.1718 \
1.618

A scalar in Linear Algebra is just a number like our bias, 1.38. The poor little thing has no direction or dimension, also called a 0-dimensional number, or a 0th-order tensor. It just sits there waiting to be called upon. Vectors on the other hand have both magnitude and direction. They got things to do! A vector could be the speed and direction of a baseball pitchers fastball. Our inputs and weights are three-dimensional vectors because they have three numbers. There's no limit to the amount of numbers a vector can contain. Typically denoted as n-dimensional.

In NumPy, a row vector can be represented as a 1-dimensional array, `np.array([3.141, 2.718, 1.618])`, and a column vector as a 2-dimensional array, `np.array([[3.141], [2.718], [1.618]])`. Inputs and weights can be 1-dimensional arrays representing vectors, but these are treated as 1D arrays in NumPy. Vectors in 3D space ([x,y,z]) are represented as 1D arrays in NumPy, though they're conceptually "3-dimensional."

Lots of things can be done to vectors: addition, subtraction, scaling, dot product. You've made it out of elementary school I assume, so you're familiar with the first two, scaling changes the magnitude/length of the vector. When referring to vectors those two words are synonymous. But scaling does not change the direction (with one exception). To scale a vector it is **multiplied** by a scalar. Note the difference, a bias shifts the value of the output by a fixed amount but doesn't scale the vector.

Scaling:
- Changes the magnitude (length) of a vector.
- Does not change the direction of the vector if the scalar is positive.
- Flips the direction of the vector if the scalar is negative (reverses direction).
- Affects every component of the vector by multiplying them by the scalar.

Example: 
- 2⋅[3,4]=[6,8] (Magnitude doubles, direction remains the same).

Example: 
- −2⋅[3,4]=[−6,−8] (Magnitude doubles, direction flips).

Bias:
- Shifts the output of a computation, but does not change the magnitude or direction of a vector.
- Adds a constant value to the weighted sum of inputs, providing an offset.
- Does not affect the vector's components directly; it only changes the output after the weighted sum.
- Used in neural networks to allow for flexibility in learning and threshold adjustments.

**Dot product** is what was used to calculate the output earlier 4.191846. Take two scalars that have the same number of elements in them, multiply each element, the add to the next two multiplied elements, rinse and repeat. It's easier to visually see what's going on vs. attempting to describe it written out. `(inputs[0]*weights[0] + inputs[1]*weights[1] + inputs[2]*weights[2])`. All of that became one number 4.191846 which is what? That's right, a scalar.

Let's use NumPy to do the exact same thing.

In [5]:
np_outputs = np.dot(inputs, weights) +  bias
print(np_outputs)

4.191846
